streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java

```diff
@@ -36,36 +36,35 @@
     protected final InternalStreamsBuilder builder;
     protected final String name;
     protected final Set<String> sourceNodes;
-    protected final StreamsGraphNode parentGraphNode;
+    protected final StreamsGraphNode streamsGraphNode;
 
     // This copy-constructor will allow to extend KStream
     // and KTable APIs with new methods without impacting the public interface.
     public AbstractStream(AbstractStream<K> stream) {
         this.builder = stream.builder;
         this.name = stream.name;
         this.sourceNodes = stream.sourceNodes;
-        this.parentGraphNode = stream.parentGraphNode;
+        this.streamsGraphNode = stream.streamsGraphNode;
     }
 
     AbstractStream(final InternalStreamsBuilder builder,
                    final String name,
                    final Set<String> sourceNodes,
-                   final StreamsGraphNode parentGraphNode) {
+                   final StreamsGraphNode streamsGraphNode) {
         if (sourceNodes == null || sourceNodes.isEmpty()) {
             throw new IllegalArgumentException("parameter <sourceNodes> must not be null or empty");
         }
 
         this.builder = builder;
         this.name = name;
         this.sourceNodes = sourceNodes;
-        this.parentGraphNode = parentGraphNode;
+        this.streamsGraphNode = streamsGraphNode;
     }
 
     protected void addGraphNode(final StreamsGraphNode newNode) {
-        //TODO remove this once actually building the topology with Graph
-        if (parentGraphNode != null) {
-            parentGraphNode.addChildNode(newNode);
-        }
+        streamsGraphNode.addChildNode(newNode);
+        builder.maybeAddNodeForOptimizationMetadata(newNode);
+
     }
 
     // This method allows to expose the InternalTopologyBuilder instance
```

Comment:
This is added for users that want to extend `AbstractStream`, one use case of it is https://github.com/fhussonnois/kafkastreams-cep.

Regarding `InternalTopologyBuilder` and `InternalStreamsBuilder`: we need to pass in `InternalTopologyBuilder` into the `StreamsGraphNode` because their `writeToTopology` (this is the one that translates the logical node into one or more physical nodes) needs it, and those graph nodes are accessed from the `InternalStreamsBuilder`, so I think we still need to let it hold a reference of the `InternalTopologyBuilder` anyways.